### DISCUSSION TITLE
Add `hermitian adjacency matrix`

### DISCUSSION THREAD

**Discussion by KeishiS** [networkx/networkx#6524] [Ideas]
I’d like to implement `hermitian adjacency matrix` and some associated functions.
Hermitian adjacency matrix has been proposed in [1] and these days the matrix is used by various methods, especially spectral clustering algorithms for digraph. e.g. [2], [3]

I'd like to access these matrices through major library. Would it be welcome?

[1] Liu, Jianxi, and Xueliang Li. ‘Hermitian-Adjacency Matrices and Hermitian Energies of Mixed Graphs’. Linear Algebra and Its Applications 466 (2015): 182–207. https://doi.org/10.1016/j.laa.2014.10.028.
[2] Cucuringu, Mihai, Huan Li, He Sun, and Luca Zanetti. ‘Hermitian Matrices for Clustering Directed Graphs: Insights and Applications’. In Proceedings of the Twenty Third International Conference on Artificial Intelligence and Statistics, edited by Silvia Chiappa and Roberto Calandra, 108:983–92. Proceedings of Machine Learning Research. PMLR, 26--28 Aug 2020. https://proceedings.mlr.press/v108/cucuringu20a.html.
[3] Laenen, Steinar, and He Sun. ‘Higher-Order Spectral Clustering of Directed Graphs’. In Advances in Neural Information Processing Systems, edited by H. Larochelle, M. Ranzato, R. Hadsell, M. F. Balcan, and H. Lin, 33:941–51. Curran Associates, Inc., 2020. https://proceedings.neurips.cc/paper/2020/file/0a5052334511e344f15ae0bfafd47a67-Paper.pdf.

---

## Definitions

In a graph $G$ with $n$ nodes, the simple hermitian adjacency matrix $A\in \mathbb{C}^{N\times N}$ of $G$ is the complex-valued matrix defined as following:

$$
\begin{aligned}
A_{u,v} = \begin{cases}
1 & u \rightsquigarrow v \quad\\&\quad v \rightsquigarrow u
\\
i & u \rightsquigarrow v \quad \\&\quad v \not\rightsquigarrow u
\\
-i & u \not\rightsquigarrow v \quad\\&\quad v \rightsquigarrow u
\\
0 & \text{o.w.}
\end{cases}
\end{aligned}
$$

```python
>>> import networkx as nx
>>> import numpy as np

>>> dg = nx.DiGraph()
>>> dg.add_edges_from([(1, 2), (1, 3), (3, 2), (1, 4), (4, 1)])
>>> nx.adjacency_matrix(dg).todense()
array([[0, 1, 1, 1],
       [0, 0, 0, 0],
       [0, 1, 0, 0],
       [1, 0, 0, 0]])
>>> np.array([[0, 1j, 1j, 1], [-1j, 0, -1j, 0], [-1j, 1j, 0, 0], [1, 0, 0, 0]]) # hermitian adjacency matrix
array([[ 0.+0.j,  0.+1.j,  0.+1.j,  1.+0.j],
       [-0.-1.j,  0.+0.j, -0.-1.j,  0.+0.j],
       [-0.-1.j,  0.+1.j,  0.+0.j,  0.+0.j],
       [ 1.+0.j,  0.+0.j,  0.+0.j,  0.+0.j]])
```

**Comment by MridulS** [comment]
Yes!!!! Please do, I would love to get this in. I vaguely remember playing around with hermitian matrices for directed graph community detection in the old days :)

**Reply by KeishiS** [reply]
Thank you for your reply!
As you mentioned, I'd like to enhance analytical tools for directed graphs. To that end, I'm thinking of implementing the basic ones first.
I'll send a pull request as soon as possible.

**Reply by KeishiS** [reply]
I'm working on this implementation, but it is taking me some time to consider edge cases, particularly in certain situations involving weighted directed graphs.

One specific edge case is the value of the hermitian adjacency matrix in situations where the weight function is denoted as `w`, and `w(u,v) != 0, w(v,u) != 0, w(u,v) != w(v,u)` .

**Reply by dschult** [reply]
Do you have a pointer to a definition of a hermitian adjacency matrix for weighted graphs? The definition given above seems to only work with unweighted graphs. And it is not obvious how to extend its definition to the weighted case. For example, if both directed edges are present the definition above says the value should be 1. But for weighted edges with different weights, a single real number cannot fully encode the two real-valued weights.

Perhaps this formulation is not intended to apply to weighted edges?

**Reply by KeishiS** [reply]
Thanks for your comment. The hermitian adjacency matrix for weighted graphs can be found, for example in [3] above and [4].
In [3], the value of the matrix is defined as $A_{u,v} = \overline{A_{v,u}} = w(u,v) \omega$ if $u\rightsquigarrow v$ where $\omega \in \mathbb{C}$ is a specific derived complex value.
In [4], the value is defined as $A_{u,v} = w(u,v) \alpha + w(v,u) \overline{\alpha}$ where $\alpha \in \mathbb{C}$ is a complex value. In these research papers, the solution for the edge case I mentioned is not explicitly stated.

Therefore, as you pointed out, the simplest implementation is to define it only for unweighted graphs.

[4] Mohar, Bojan. “A new kind of Hermitian matrices for digraphs.” Linear Algebra and its Applications (2019): n. pag.

**Reply by dschult** [reply]
Hmmm... It looks like at least the definition you give from [4] would help with your edge case so long as the edge weights are real values (which I guess they are).  A_{u,v} = w(u,v) * alpha + w(v,u) * alphabar has complex conjugate w(u, v) * alphabar + w(v,u) * alpha, which **is** the value given by the original formula applied to A_{v,u}

So in your case where w(v,u) and w(u,v) are not zero and not equal you should set A_{u,v} and A_{v,u} according to that same formula and it should work... Though maybe I am missing something...

**Reply by KeishiS** [reply]
I'm sorry, I was confused. Your understanding is correct. The definition from [4] is useful in the edge case I mentioned above.

The point I'd like to discuss is that there are at least two definitions of the hermitian adjacency matrix, and these two don't coincide (the definition from [4] is not a generalization of the definition from [3]). While there is no problem with the definition from [4], the definition from [3] does not provide a clear correspondence for weighted graphs.

Is it ok to switch the returned value based on a parameter passed as arguments and to raise an exception when a weighted graph is passed with a specific parameter?
e.g.
`hermitian_adjacency_matrix(g, method="defi[4]") => unweighted/weighted OK`
`hermitian_adjacency_matrix(g, method="defi[3]") => unweighted OK, weighted NG`

**Comment by dschult** [comment]
I suspect the description in [3] is a typo -- and that it does actually work for weighted graphs. We just need to figure out what it should be. We could figure that out by looking at papers (perhaps by the same authors and more hopefully more recently) which describe that formula -- then we can check whether they agree. If not papers by them, then papers they cite -- or papers that cite them.  Once we get tired of that approach, we could contact them and ask for clarification since their formula doesn't work for weighted graphs.  Still -- even when we figure out the typo, it is pretty clear that the two papers are using different definitions of how to encode a directed graph in a hermitian matrix. 

That problematic feature is one of the reasons we don't already have a hermitian adjacency matrix function. The literature hasn't matured enough to figure out which formulation is the most useful. It would help a lot if researcher could find an encoding that leads to spectral meaning from the matrix. That's presumably the whole reason for getting a hermitian matrix -- so that the eigenvalues are real numbers. An asymmetric real matrix is a lot simpler than a symmetric complex-valued matrix for many tasks other than spectral features.

Do **you** have an application or problem that involves these hermitian adjacency matrices? What is driving your interest in them? That might be a good way to see how to move forward.

**Reply by KeishiS** [reply]
My motivation is to explore methodologies for extracting global structures in directed graphs, such as the existence of communities and the relationships between communities. I believe that focusing on the eigenvalues and eigenvectors of hermitian matrices as a representation could be useful for this purpose. I am also in the midst of conducting research in this area and do not have a clear answer as to which representation is most useful. It may indeed have been premature to proceed with the implementation under such circumstances.

Starting with the literature survey you advised, I'd like to prepare convincing answers by contacting other researchers and advancing my own research. Of course, if someone else is able to take on this challenge first, it would be a wonderful thing.
